src/trace/connection.rs

```diff
@@ -0,0 +1,188 @@
+            reason = event.reason.tracing_representation(),
+            "Connection closed",
+        );
+    }
+
+    fn handle_connection_checkout_started_event(&self, event: ConnectionCheckoutStartedEvent) {
+        tracing::debug!(
+            target: CONNECTION_TRACING_EVENT_TARGET,
+            topologyId = self.topology_id.tracing_representation(),
+            serverHost = event.address.host(),
+            serverPort = event.address.port(),
+            "Connection checkout started",
+        );
+    }
+
+    fn handle_connection_checkout_failed_event(&self, event: ConnectionCheckoutFailedEvent) {
+        tracing::debug!(
+            target: CONNECTION_TRACING_EVENT_TARGET,
+            topologyId = self.topology_id.tracing_representation(),
+            serverHost = event.address.host(),
+            serverPort = event.address.port(),
+            reason = event.reason.tracing_representation(),
+            "Connection checkout failed",
+        );
+    }
+
+    fn handle_connection_checked_out_event(&self, event: ConnectionCheckedOutEvent) {
+        tracing::debug!(
+            target: CONNECTION_TRACING_EVENT_TARGET,
+            topologyId = self.topology_id.tracing_representation(),
+            serverHost = event.address.host(),
+            serverPort = event.address.port(),
+            driverConnectionId = event.connection_id,
+            "Connection checked out",
+        );
+    }
+
+    fn handle_connection_checked_in_event(&self, event: ConnectionCheckedInEvent) {
+        tracing::debug!(
+            target: CONNECTION_TRACING_EVENT_TARGET,
+            topologyId = self.topology_id.tracing_representation(),
+            serverHost = event.address.host(),
+            serverPort = event.address.port(),
+            driverConnectionId = event.connection_id,
+            "Connection checked in",
+        );
+    }
+}
+
+impl TracingRepresentation for ConnectionClosedReason {
+    type Representation = &'static str;
+
+    fn tracing_representation(self) -> &'static str {
+        match self {
+            ConnectionClosedReason::Stale => "Connection became stale because the pool was cleared",
+            ConnectionClosedReason::Idle => {
+                "Connection has been available but unused for longer than the configured max idle \
+                 time"
+            }
+            ConnectionClosedReason::Error => "An error occurred while using the connection",
+            ConnectionClosedReason::Dropped => "Connection was dropped during an operation",
+            ConnectionClosedReason::PoolClosed => "Connection pool was closed",
+        }
+    }
+}
+
+impl TracingRepresentation for ConnectionCheckoutFailedReason {
+    type Representation = &'static str;
+
+    fn tracing_representation(self) -> &'static str {
+        match self {
+            ConnectionCheckoutFailedReason::Timeout => {
+                "Wait queue timeout elapsed without a connection becoming available"
```

Comment:
We don't actually use this reason anywhere in the driver because we don't have a wait queue timeout, but I think once DRIVERS-2347 is done in Rust, we should start using this for connection establishment timeout. To prepare for that, we might want to update this string accordingly.

Reply:
Just to check my understanding from reading over that ticket, would this specifically mean that we hit `connectTimeoutMS`? e.g. should the message be something like "Failed to establish a new connection within connectTimeoutMS"?

Reply:
Yep that sounds perfect.